File: internal/agent/advisors/runner.go
package advisors

import (
	"context"
	"fmt"
	"strings"
	"sync"
	"time"

	"github.com/nebolabs/nebo/internal/agent/ai"
	"github.com/nebolabs/nebo/internal/agent/session"
)

// MaxAdvisors is the maximum number of advisors that can run concurrently.
// More voices = noise, not intelligence.
const MaxAdvisors = 5

// AdvisorTimeout is the maximum time to wait for all advisors to respond.
const AdvisorTimeout = 30 * time.Second

// Runner executes advisors in parallel and collects their responses.
// Advisors never talk to each other - they talk TO the main agent.
type Runner struct {
	loader   *Loader
	provider ai.Provider
}

// NewRunner creates an advisor runner
func NewRunner(loader *Loader, provider ai.Provider) *Runner {
	return &Runner{
		loader:   loader,
		provider: provider,
	}
}

// SetProvider updates the provider used for advisor execution
func (r *Runner) SetProvider(provider ai.Provider) {
	r.provider = provider
}

// Deliberate runs all enabled advisors in parallel on the given task.
// Returns a slice of responses (may be partial if some advisors timeout/fail).
// The task should be a clear description of what the main agent is about to do.
func (r *Runner) Deliberate(ctx context.Context, task string, recentMessages []session.Message) ([]Response, error) {
	advisors := r.loader.List()
	if len(advisors) == 0 {
		return nil, nil // No advisors configured
	}

	// Cap at MaxAdvisors
	if len(advisors) > MaxAdvisors {
		advisors = advisors[:MaxAdvisors]
	}

	fmt.Printf("[advisors] Starting deliberation with %d advisors for task: %s\n",
		len(advisors), truncateTask(task, 100))

	// Use max per-advisor timeout as the overall deadline
	maxTimeout := AdvisorTimeout
	for _, adv := range advisors {
		if adv.TimeoutSeconds > 0 {
			t := time.Duration(adv.TimeoutSeconds) * time.Second
			if t > maxTimeout {
				maxTimeout = t
			}
		}
	}
	ctx, cancel := context.WithTimeout(ctx, maxTimeout)
	defer cancel()

	// Run advisors in parallel
	var wg sync.WaitGroup
	responses := make([]Response, len(advisors))
	errors := make([]error, len(advisors))

	for i, advisor := range advisors {
		wg.Add(1)
		go func(idx int, adv *Advisor) {
			defer wg.Done()
			// Per-advisor timeout
			advCtx := ctx
			if adv.TimeoutSeconds > 0 {
				var advCancel context.CancelFunc
				advCtx, advCancel = context.WithTimeout(ctx, time.Duration(adv.TimeoutSeconds)*time.Second)
				defer advCancel()
			}
			resp, err := r.runAdvisor(advCtx, adv, task, recentMessages)
			if err != nil {
				errors[idx] = err
				fmt.Printf("[advisors] Advisor %s failed: %v\n", adv.Name, err)
			} else {
				responses[idx] = resp
				fmt.Printf("[advisors] Advisor %s responded (confidence: %d)\n",
					adv.Name, resp.Confidence)
			}
		}(i, advisor)
	}

	// Wait for all advisors to complete
	wg.Wait()

	// Collect successful responses
	var results []Response
	for i, resp := range responses {
		if errors[i] == nil && resp.Critique != "" {
			results = append(results, resp)
		}
	}

	fmt.Printf("[advisors] Deliberation complete: %d/%d advisors responded\n",
		len(results), len(advisors))

	return results, nil
}

// runAdvisor executes a single advisor and parses its response
func (r *Runner) runAdvisor(ctx context.Context, advisor *Advisor, task string, recentMessages []session.Message) (Response, error) {
	resp := Response{
		AdvisorName: advisor.Name,
		Role:        advisor.Role,
	}

	// Build the prompt for this advisor
	systemPrompt := advisor.BuildSystemPrompt(task)

	// Build messages: include recent context + the task
	messages := buildAdvisorMessages(recentMessages, task)

	// Call the provider
	events, err := r.provider.Stream(ctx, &ai.ChatRequest{
		System:   systemPrompt,
		Messages: messages,
	})
	if err != nil {
		return resp, fmt.Errorf("stream error: %w", err)
	}

	// Collect response
	var content strings.Builder
	for event := range events {
		switch event.Type {
		case ai.EventTypeText:
			content.WriteString(event.Text)
		case ai.EventTypeError:
			return resp, event.Error
		}
	}

	// Parse the response into structured format
	resp.Critique = content.String()
	resp.Confidence = extractConfidence(content.String())
	resp.Risks = extractSection(content.String(), "Risks")
	resp.Suggestion = extractSection(content.String(), "Suggestion")

	return resp, nil
}

// buildAdvisorMessages creates the message list for an advisor call
func buildAdvisorMessages(recentMessages []session.Message, task string) []session.Message {
	// Include a summary of recent context (last 3-5 messages for brevity)
	maxContext := 5
	if len(recentMessages) > maxContext {
		recentMessages = recentMessages[len(recentMessages)-maxContext:]
	}

	// Build context summary
	var contextSummary strings.Builder
	contextSummary.WriteString("Recent conversation context:\n\n")

	for _, msg := range recentMessages {
		switch msg.Role {
		case "user":
			contextSummary.WriteString(fmt.Sprintf("User: %s\n\n", truncateTask(msg.Content, 200)))
		case "assistant":
			if msg.Content != "" {
				contextSummary.WriteString(fmt.Sprintf("Assistant: %s\n\n", truncateTask(msg.Content, 200)))
			}
		}
	}

	return []session.Message{
		{
			Role:    "user",
			Content: contextSummary.String() + "\n---\n\nThe main agent is about to respond. Please provide your critique.",
		},
	}
}

// FormatForInjection formats advisor responses for injection into the system prompt
func FormatForInjection(responses []Response) string {
	if len(responses) == 0 {
		return ""
	}

	var sb strings.Builder
	sb.WriteString("\n\n---\n## Internal Deliberation (Advisor Perspectives)\n\n")
	sb.WriteString("Before responding, consider these internal perspectives:\n\n")

	for _, resp := range responses {
		sb.WriteString(fmt.Sprintf("### %s (%s)\n", resp.AdvisorName, resp.Role))
		sb.WriteString(resp.Critique)
		sb.WriteString("\n\n")
	}

	sb.WriteString("---\n\n")
	sb.WriteString("Synthesize these perspectives but make your own decision. You are the authority.\n")

	return sb.String()
}

// extractConfidence attempts to extract a confidence score (1-10) from the response
func extractConfidence(text string) int {
	// Look for patterns like "Confidence: 7" or "7/10"
	lines := strings.Split(text, "\n")
	for _, line := range lines {
		lower := strings.ToLower(line)
		if strings.Contains(lower, "confidence") {
			// Try to find a number
			for _, word := range strings.Fields(line) {
				word = strings.Trim(word, "():,*/")
				if len(word) == 1 && word[0] >= '1' && word[0] <= '9' {
					return int(word[0] - '0')
				}
				if len(word) == 2 && word[0] == '1' && word[1] == '0' {
					return 10
				}
			}
		}
	}
	return 5 // Default confidence if not found
}

// extractSection attempts to extract a named section from the response
func extractSection(text, sectionName string) string {
	lines := strings.Split(text, "\n")
	inSection := false
	var result strings.Builder

	for _, line := range lines {
		lower := strings.ToLower(line)
		if strings.Contains(lower, strings.ToLower(sectionName)) && strings.Contains(line, ":") {
			inSection = true
			// Get content after the colon on the same line
			parts := strings.SplitN(line, ":", 2)
			if len(parts) > 1 {
				content := strings.TrimSpace(parts[1])
				if content != "" {
					return content
				}
			}
			continue
		}

		if inSection {
			// Stop at next section header or empty line after content
			if strings.HasPrefix(line, "#") || strings.HasPrefix(line, "**") {
				break
			}
			if line == "" && result.Len() > 0 {
				break
			}
			if line != "" {
				result.WriteString(line)
				result.WriteString(" ")
			}
		}
	}

	return strings.TrimSpace(result.String())
}

// truncateTask truncates a task description to a maximum length
func truncateTask(task string, maxLen int) string {
	if len(task) <= maxLen {
		return task
	}
	return task[:maxLen-3] + "..."
}

File: crates/agent/src/advisors/runner.rs
use std::sync::Arc;

use futures::stream::FuturesUnordered;
use futures::StreamExt;
use tracing::{debug, warn};

use ai::{ChatRequest, Message, Provider, StreamEventType};

use super::advisor::{Advisor, Response};
use super::loader::Loader;

/// Maximum number of advisors to consult in parallel.
const MAX_ADVISORS: usize = 5;

/// Default per-advisor timeout in seconds.
const DEFAULT_TIMEOUT_SECS: u64 = 30;

/// Runs advisor deliberation: consults enabled advisors in parallel via LLM.
pub struct Runner {
    loader: Arc<Loader>,
    providers: Arc<Vec<Arc<dyn Provider>>>,
}

impl Runner {
    pub fn new(loader: Arc<Loader>, providers: Arc<Vec<Arc<dyn Provider>>>) -> Self {
        Self { loader, providers }
    }

    /// Run deliberation: query up to MAX_ADVISORS enabled advisors in parallel,
    /// each with their own system prompt, and aggregate responses.
    pub async fn deliberate(&self, task: &str) -> Result<Vec<Response>, String> {
        let provider = self
            .providers
            .first()
            .ok_or_else(|| "no AI providers configured".to_string())?;

        let advisors = self.loader.list_enabled().await;
        if advisors.is_empty() {
            return Ok(Vec::new());
        }

        let capped: Vec<Advisor> = advisors.into_iter().take(MAX_ADVISORS).collect();
        debug!(count = capped.len(), task = %task, "starting advisor deliberation");

        let mut futures = FuturesUnordered::new();

        for advisor in &capped {
            let advisor = advisor.clone();
            let task = task.to_string();
            let timeout_secs = if advisor.timeout_seconds > 0 {
                advisor.timeout_seconds as u64
            } else {
                DEFAULT_TIMEOUT_SECS
            };

            let provider_ref: &dyn Provider = &**provider;

            futures.push(async move {
                let result = tokio::time::timeout(
                    std::time::Duration::from_secs(timeout_secs),
                    run_single_advisor(provider_ref, &advisor, &task),
                )
                .await;

                match result {
                    Ok(Ok(response)) => Some(response),
                    Ok(Err(e)) => {
                        warn!(advisor = %advisor.name, error = %e, "advisor deliberation failed");
                        None
                    }
                    Err(_) => {
                        warn!(advisor = %advisor.name, timeout_secs, "advisor timed out");
                        None
                    }
                }
            });
        }

        let mut responses = Vec::new();
        while let Some(result) = futures.next().await {
            if let Some(response) = result {
                responses.push(response);
            }
        }

        // Sort by confidence (highest first)
        responses.sort_by(|a, b| b.confidence.cmp(&a.confidence));

        debug!(
            count = responses.len(),
            "advisor deliberation complete"
        );

        Ok(responses)
    }

    /// Format advisor responses for injection into the main agent's context.
    pub fn format_for_injection(responses: &[Response]) -> String {
        if responses.is_empty() {
            return String::new();
        }

        let mut output = String::from(
            "---\n## Internal Deliberation (Advisor Perspectives)\n\n\
             Before responding, consider these internal perspectives:\n\n",
        );

        for response in responses {
            output.push_str(&format!("### {} ({})\n", response.advisor_name, response.role));
            output.push_str(&response.critique);
            if !response.risks.is_empty() {
                output.push_str(&format!("\n**Risks:** {}", response.risks));
            }
            if !response.suggestion.is_empty() {
                output.push_str(&format!("\n**Suggestion:** {}", response.suggestion));
            }
            output.push_str(&format!("\n*Confidence: {}/10*\n\n", response.confidence));
        }

        output.push_str("---\n\nSynthesize these perspectives but make your own decision. You are the authority.\n");

        output
    }
}

/// Run a single advisor's LLM call and parse the response.
async fn run_single_advisor(
    provider: &dyn Provider,
    advisor: &Advisor,
    task: &str,
) -> Result<Response, String> {
    let system_prompt = advisor.build_system_prompt(task);

    let req = ChatRequest {
        messages: vec![Message {
            role: "user".to_string(),
            content: format!(
                "Analyze this task and provide your perspective:\n\n{}",
                task
            ),
            ..Default::default()
        }],
        tools: vec![],
        max_tokens: 1024,
        temperature: 0.7,
        system: system_prompt,
        static_system: String::new(),
        model: String::new(),
        enable_thinking: false,
        metadata: None,
        cache_breakpoints: vec![],
        cancel_token: None,
    };

    let mut rx = provider
        .stream(&req)
        .await
        .map_err(|e| format!("provider error: {}", e))?;

    let mut full_text = String::new();
    while let Some(event) = rx.recv().await {
        match event.event_type {
            StreamEventType::Text => {
                full_text.push_str(&event.text);
            }
            StreamEventType::Error => {
                return Err(event.error.unwrap_or_else(|| "unknown error".into()));
            }
            StreamEventType::Done => break,
            _ => {}
        }
    }

    if full_text.is_empty() {
        return Err("empty response from advisor".into());
    }

    let confidence = Response::extract_confidence(&full_text);
    let risks = Response::extract_section(&full_text, "Risks");
    let suggestion = Response::extract_section(&full_text, "Suggestion");
    let assessment = Response::extract_section(&full_text, "Assessment");

    let critique = if assessment.is_empty() {
        // If no structured assessment, use the full text
        full_text.clone()
    } else {
        assessment
    };

    Ok(Response {
        advisor_name: advisor.name.clone(),
        role: advisor.role.clone(),
        critique,
        confidence,
        risks,
        suggestion,
    })
}

/// Implement the AdvisorDeliberator trait from tools crate
/// so that AgentTool can call deliberate without circular dependencies.
impl tools::bot_tool::AdvisorDeliberator for Runner {
    fn deliberate<'a>(
        &'a self,
        task: &'a str,
    ) -> std::pin::Pin<Box<dyn std::future::Future<Output = Result<String, String>> + Send + 'a>>
    {
        Box::pin(async move {
            let responses = self.deliberate(task).await?;
            Ok(Self::format_for_injection(&responses))
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_format_for_injection_empty() {
        let output = Runner::format_for_injection(&[]);
        assert!(output.is_empty());
    }

    #[test]
    fn test_format_for_injection() {
        let responses = vec![
            Response {
                advisor_name: "skeptic".to_string(),
                role: "critic".to_string(),
                critique: "This approach has significant risks.".to_string(),
                confidence: 8,
                risks: "Could fail under load.".to_string(),
                suggestion: "Add caching first.".to_string(),
            },
            Response {
                advisor_name: "pragmatist".to_string(),
                role: "builder".to_string(),
                critique: "The approach is practical and achievable.".to_string(),
                confidence: 7,
                risks: String::new(),
                suggestion: "Ship the MVP.".to_string(),
            },
        ];

        let output = Runner::format_for_injection(&responses);
        assert!(output.contains("### skeptic (critic)"));
        assert!(output.contains("### pragmatist (builder)"));
        assert!(output.contains("significant risks"));
        assert!(output.contains("Confidence: 8/10"));
        assert!(output.contains("Synthesize these perspectives"));
    }
}
